File: BaekJoon Online Judge/C++/Step14-Math_3 Number Theory and Combinatorics/2981.cpp
//
//  2981.cpp
//  Step14-Math_3 Number Theory and Combinatorics
//
//  Created by 김예빈 on 2019. 10. 1..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <algorithm>
using namespace std;

int num[101], result[500];

int gcd(int, int);

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, max, cnt = 0;
    
    cin >> n;
    
    for(int i = 0; i < n; i++)
        cin >> num[i];
    
    sort(num, num + n);
    max = num[1] - num[0];
    
    for(int i = 2; i < n; i++)
        max = gcd(max, num[i] - num[i - 1]);
    
    for(int i = 1; i*i <= max; i++)
        if(max % i == 0) {
            result[cnt++] = i;
            
            if(i != max / i)
                result[cnt++] = max / i;
        }
    
    sort(result, result + cnt);
    
    for(int i = 0; i < cnt; i++) {
        if(result[i] != 1) {
            cout << result[i];
            
            if(i < cnt - 1)
                cout << " ";
        }
    }
    
    return 0;
}

int gcd(int a, int b) {
    if(a % b == 0)
        return b;
    else
        return gcd(b, a % b);
}

File: BaekJoon Online Judge/Swift/Step16-Math_3 Number Theory and Combinatorics/2981.swift
//
//  2981.swift
//  Step16-Math_3 Number Theory and Combinatorics
//
//  Created by Yebin Kim on 2019/12/25.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

func euclidean(_ a: Int, _ b: Int) -> Int {
    if b % a == 0 {
        return a
    } else {
        return euclidean(b % a, a)
    }
}

if let N = Int(readLine()!) {
    var num = [Int]()
    var answer = [Int]()

    for _ in 0..<N {
        num.append(Int(readLine()!)!)
    }

    num.sort()
    var gcd = num[1] - num[0]

    for i in 2..<num.count {
        gcd = euclidean(gcd, num[i] - num[i - 1])
    }
    
    var i = 1
    
    while i * i <= gcd {
        if gcd % i == 0 {
            answer.append(i)
            
            if i != gcd / i {
                answer.append(gcd / i)
            }
        }
        
        i += 1
    }
    
    answer = answer.filter { $0 != 1 }
    answer.sort()
    
    for (_, e) in answer.enumerated() {
        print(e, terminator: " ")
    }
}
